src/js/components/OptionalEnviromentComponent.jsx

```diff
@@ -0,0 +1,99 @@
+var React = require("react/addons");
+
+var appValidator = require("../validators/appValidator");
+var FormGroupComponent = require("../components/FormGroupComponent");
+var DuplicableRowControls = require("../components/DuplicableRowControls");
+
+var OptionalSettingsComponent = React.createClass({
+  displayName: "OptionalEnvironmentComponent",
+
+  propTypes: {
+    errors: React.PropTypes.array,
+    model: React.PropTypes.object.isRequired
+  },
+
+  getInitialState: function () {
+    var env = this.props.model.env;
+    var state = {
+      env: Object.keys(env).map(function (key) {
+        return {
+          key: key,
+          value: env[key]
+        };
+      })
+    };
+    if (state.env.length === 0) {
+      state.env = [true];
+    }
+    return state;
+  },
+
+  handleAddRow: function (position, event) {
+    event.target.blur();
+    event.preventDefault();
+    var env = this.state.env.concat(true);
+    this.setState({env: env});
+  },
+
+  handleRemoveRow: function (position, event) {
+    event.target.blur();
+    event.preventDefault();
+    var env = this.state.env.slice();
+    env = env.map(function (value, index) {
+      return value && index !== position;
+    });
+    if (env.filter((exists) => exists).length === 0) {
+      env.push(true);
+    }
```

Comment:
Again, this `true` flag is kind of obscure as to what it means at the end of an array of objects. Have you thought about extracting it to its own state?